File: handwrittenDigits.m
% NOTE Run the following code on a production machine,
%   taking less than 8 hours

%% neural network model for handwritten digits

%   Train handwritten digits using neural network model
%   Data from MNIST, 5000 samples, each 400 features, classify to 10 classes [1, 10]
%   Using architecture [400, 25, 10], input layer 400 elements; 1 hidden layer with 25 eflements; output layer 10 elements
%   Run 3 times for optmization. Run 1: optimize sample number; Run 2:
%   optimize regulating rate; Run 3: compute Theta and test on test set

%% Load data
close all;
clear;
clc;

addpath('./lib/');
addpath('./data_handwritten_digits/');

load('handwrittenDigits.mat');
if ~exist('X', 'var') || ~exist('y', 'var')
    fprintf('ERR, X or y not contained in mat file\n');
    return;
end
XRaw = X;
yRaw = y;

%% Preperation
%   shuffle data set
%   divide into training set (60%), cross validation set (20%) and test set (20%)
%   initial parameters

[XTotal, yTotal] = shuffle(XRaw, yRaw);

[X, y, XCV, yCV, XTest, yTest] = split(XTotal, yTotal);

[sampleNumber, featureNumber] = size(X);
classNumber = 10;
architecturePara = [400, 25, 10];
maxIter = 400;
ThetaCellInitial = randThetaCell(architecturePara);

%% Optimize sample number (takes < 5h)
sampleNumberOpt = sampleNumberVsCost(X, y, XCV, yCV, ThetaCellInitial, 0, maxIter, architecturePara, 2950:1:3000);
X = X(1:sampleNumberOpt, :);
y = y(1:sampleNumberOpt, :);

%% Optimize regulating rate (lambda)
regulatingRateOpt = regulatingRateVsCost(X, y, XCV, yCV, ThetaCellInitial, 0:1:10, maxIter, architecturePara);
regulatingRate = regulatingRateOpt;

%% Compute Theta
maxIter = 4000;
ThetaCell = train(X, y, ThetaCellInitial, regulatingRate, maxIter, architecturePara);

%% Accuracy
accuracyTrain = accuracy(predict(ThetaCell, X), y);
accuracyCV = accuracy(predict(ThetaCell, XCV), yCV);
accuracyTest = accuracy(predict(ThetaCell, XTest), yTest);
fprintf('train accuracy %.2f\n', accuracyTrain);
fprintf('cross validation accuracy %.2f\n', accuracyCV);
fprintf('test accuracy %.2f\n', accuracyTest);

%% Clean
rmpath('./lib/');
rmpath('./data_handwritten_digits/');

File: handwritten_digits.py
'''\
	train neural network model using handwritten digits data
	python handwritten_digits.py
'''

from sklearn.utils import shuffle
import time
import datah
import stor
import nnp

def prepare_data():
    # run once
    X_path = './data_handwritten_digits/X_total.csv'
    y_path = './data_handwritten_digits/y_total.csv'
    X_total, y_total = datah.Data(X_path).to_nparray(float), datah.Data(y_path).to_nparray(int)
    X_total, y_total = nnp.regulate(X_total), nnp.regulate(y_total)
    for i in range(0, y_total.shape[0]):
        y_total[i] = y_total[i] if y_total[i] != 10 else 0
    X_total, y_total = shuffle(X_total, y_total)
    X_train, y_train = X_total[:3000], y_total[:3000]
    X_cv, y_cv = X_total[3000:4000], y_total[3000:4000]
    X_test, y_test = X_total[4000:], y_total[4000:]
    # save data
    box = stor.Box('./data_handwritten_digits/')
    box.put('X_total', X_total).put('y_total', y_total)\
        .put('X_train', X_train).put('y_train', y_train)\
        .put('X_cv', X_cv).put('y_cv', y_cv)\
        .put('X_test', X_test).put('y_test', y_test)

def main():
    # get prepared data
	box = stor.Box('./data_handwritten_digits/')
	X_train, y_train = box.get('X_train'), box.get('y_train')
	X_cv, y_cv = box.get('X_cv'), box.get('y_cv')
	X_test, y_test = box.get('X_test'), box.get('y_test')

	# tune
	num_sample_opt = nnp.opt_num_sample(X_train, y_train, X_cv, y_cv, (2950, 3000))
	X_train, y_train = X_train[:num_sample_opt], y_train[:num_sample_opt]
	regulating_rate_opt = nnp.opt_regulating_rate(X_train, y_train, X_cv, y_cv, (0, 1))

	# train
	mlps = nnp.train(X_train, y_train, regulating_rate_opt)
	box.update('mlps', mlps)

	# predict y_hypo
	y_test_hypo = nnp.predict(mlps, X_test)

	# judge
	precision = nnp.judge(y = y_test, y_hypo = y_test_hypo)
	print('num_sample_opt: {}'.format(num_sample_opt))
	print('regulating_rate_opt: {}'.format(regulating_rate_opt))
	print('precision: {}'.format(precision))

if __name__ == '__main__': main()
